lib/core/src/authentication_client_utils.cpp

```diff
@@ -0,0 +1,128 @@
+        }
+        std::string password;
+        std::getline(std::cin, password);
+        fmt::print("\n");
+        tty.c_lflag = oldflag;
+        if (0 != tcsetattr(STDIN_FILENO, TCSANOW, &tty)) {
+            fmt::print("Error reinstating echo mode.\n");
+        }
+        return password;
+    } // get_password_from_client_stdin
+
+    auto read_session_token_from_file() -> std::string
+    {
+        const auto session_token_file_path = get_session_token_file_path();
+        if (!std::filesystem::exists(session_token_file_path)) {
+            return {};
+        }
+
+        std::ifstream session_token_file_stream{session_token_file_path.c_str()};
+        if (!session_token_file_stream.is_open()) {
+            const auto irods_error_code = UNIX_FILE_OPEN_ERR - errno;
+            THROW(irods_error_code,
+                  fmt::format(
+                      "Failed to open session token file [{}]. errno:[{}]", session_token_file_path.c_str(), errno));
+        }
+
+        std::string session_token_file_contents;
+        session_token_file_stream >> session_token_file_contents;
+        session_token_file_stream.close();
+        if (session_token_file_contents.size() != session_token_length) {
+            THROW(SYS_INVALID_INPUT_PARAM,
+                  fmt::format("Session token must be a 36-character UUID, but had length [{}].",
+                              session_token_file_contents.size()));
+        }
+        return session_token_file_contents;
+    } // read_session_token_from_file
+
+    auto write_session_token_to_file(const std::string& _session_token) -> void
+    {
+        if (_session_token.size() != session_token_length) {
+            THROW(
+                SYS_INVALID_INPUT_PARAM,
+                fmt::format("Session token must be a 36-character UUID, but had length [{}].", _session_token.size()));
+        }
+
+        const auto session_token_file_path = get_session_token_file_path();
+        std::ofstream session_token_file_stream{session_token_file_path.c_str()};
+        if (!session_token_file_stream.is_open()) {
+            const auto irods_error_code = UNIX_FILE_OPEN_ERR - errno;
+            THROW(irods_error_code,
+                  fmt::format(
+                      "Failed to open session token file [{}]. errno:[{}]", session_token_file_path.c_str(), errno));
+        }
+
+        session_token_file_stream << _session_token;
+        session_token_file_stream.close();
+        // Make sure that only the owner can read/write this file because it contains sensitive information.
+        std::filesystem::permissions(
+            session_token_file_path.c_str(), std::filesystem::perms::owner_read | std::filesystem::perms::owner_write);
+    } // write_session_token_to_file
+
+    auto remove_session_token_file() -> int
+    {
+        try {
+            if (!std::filesystem::remove(get_session_token_file_path())) {
+                return static_cast<int>(USER_FILE_DOES_NOT_EXIST);
```

Comment:
Is the error code guaranteed to be an existence error? What if it's a permission issue?

Reply:
If it failed to remove the file due to a permission error, a `std::filesystem::filesystem_error` is thrown. The return value for this call is, according to [`std::filesystem::remove`'s cppreference page](https://en.cppreference.com/w/cpp/filesystem/remove.html)...

> true if the file was deleted, false if it did not exist. The overload that takes error_code& argument returns false on errors.